### PR TITLE
HIVE-26754 - Implement array_distinct UDF to return an array after re…

### PR DESCRIPTION
**Pull request by tarak271**
…moving duplicates in it

### What changes were proposed in this pull request?

Implement array_distinct function in Hive

### Why are the changes needed?

This enhancement is already implemented in Spark

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Created Junit tests as well as qtests as part of this change

### CODE REVIEW

**Review by SourabhBadhya — APPROVED** [review]
LGTM +1

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    public Object evaluate(DeferredObject[] arguments) throws HiveException {
+
+        Object array = arguments[ARRAY_IDX].get();
+
+        if (arrayOI.getListLength(array) <= 0) {
+            return null;
+        }
+
+        List<?> retArray = ((ListObjectInspector) argumentOIs[ARRAY_IDX]).getList(array);
+        return convertArray(retArray.stream().distinct().collect(Collectors.toList()));
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java:57]: Do we need the ``convertArray`` method? Can we not do something with the stream like adding a `map`, ``.map(o -> converter.convert(o))``

Give a check to:
```
        return retArray.stream().distinct().map(o -> converter.convert(o)).collect(Collectors.toList());
```

**Review by ayushtkn — COMMENTED** [review]
Thanx @tarak271, Had a quick look, have dropped some comments. Please give a  check

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Generic UDF for distinct array
+ * <code>ARRAY_DISTINCT(array(obj1, obj2, obj3...))</code>.
+ *
+ * @see org.apache.hadoop.hive.ql.udf.generic.GenericUDF
+ */
+@Description(name = "array_distinct",
+        value = "_FUNC_(array(obj1, obj2,...)) - "
+                + "The function returns an array of the same type as the input argument where all duplicate"
+                + " values have been removed.",
```
[the remark this answers is not in the document]
Comment (tarak271) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java]: Changed description as recommended

File: ql/src/test/queries/clientpositive/udf_array_distinct.q [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+SELECT array_distinct(array(null)) FROM src tablesample (1 rows);
+
+SELECT array_distinct(array(1.12, 2.23, 3.34, null,1.11,1.12,2.9)) FROM src tablesample (1 rows);
+
+SELECT array_distinct(array(1.1234567890, 2.234567890, 3.34567890, null, 3.3456789, 2.234567,1.1234567890)) FROM src tablesample (1 rows);
+
+SELECT array_distinct(array(11234567890, 2234567890, 334567890, null, 11234567890, 2234567890, 334567890, null)) FROM src tablesample (1 rows);
+
+SELECT array_distinct(array(array("a","b","c","d"),array("a","b","c","d"),array("a","b","c","d","e"),null,array("e","a","b","c","d"))) FROM src tablesample (1 rows);
```
[the remark this answers is not in the document]
Comment (tarak271) [reply ql/src/test/queries/clientpositive/udf_array_distinct.q]: Added another table that contains NULL values in array, since `SELECT array_distinct(null) FROM src tablesample (1 rows);` is throwing exception from UDF code

`Caused by: org.apache.hadoop.hive.ql.exec.UDFArgumentTypeException: "array" expected at function ARRAY_DISTINCT, but "void" is found`	
`at org.apache.hadoop.hive.ql.udf.generic.AbstractGenericUDFArrayBase.checkArgCategory(AbstractGenericUDFArrayBase.`

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java (hunk 1 above)
Comment (tarak271) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java]: Made changes as recommended

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public Object evaluate(DeferredObject[] arguments) throws HiveException {
+
+        Object array = arguments[ARRAY_IDX].get();
+
+        if (arrayOI.getListLength(array) <= 0) {
+            return new ArrayList<Object>();
+        }
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java:54]: 0 denotes an empty array, but -1 denotes the argument to the UDF was ``null``, if argument is ``null``, we should return ``null`` only, if the array is empty then we should return an empty array.
Better to use ``Collections.emptyArray()``
Something like this might do:
```

        // If the array is empty, then there are no duplicates, return back the empty array
        if (arrayOI.getListLength(array) == 0) {
            return Collections.emptyList();
        } else if (arrayOI.getListLength(array) < 0) {
            return null;
        }
```

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/AbstractGenericUDFArrayBase.java [hunk 5]
```diff
@@ -106,4 +106,4 @@ ObjectInspector initListOI(ObjectInspector[] arguments) {
         return ObjectInspectorFactory.getStandardListObjectInspector(initOI(arguments));
     }
 
-}
+}
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/AbstractGenericUDFArrayBase.java:109]: unrelated change, revert

File: ql/src/test/queries/clientpositive/udf_array_distinct.q [hunk 6]
```diff
@@ -0,0 +1,33 @@
+--! qt:dataset:src
+set hive.fetch.task.conversion=more;
```
Comment (ayushtkn) [inline ql/src/test/queries/clientpositive/udf_array_distinct.q:2]: There are too many select *,
better to add
``-- SORT_QUERY_RESULTS`` to avoid any flakiness

File: ql/src/test/queries/clientpositive/udf_array_distinct.q [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+SELECT array_distinct(array(array("a","b","c","d"),array("a","b","c","d"),array("a","b","c","d","e"),null,array("e","a","b","c","d"))) FROM src tablesample (1 rows);
+
+# handle null array cases
+
+dfs ${system:test.dfs.mkdir} ${system:test.tmp.dir}/test_null_array;
+
+dfs -copyFromLocal ../../data/files/test_null_array.csv ${system:test.tmp.dir}/test_null_array/;
+
+create external table test_null_array (id int, value Array<String>) ROW FORMAT DELIMITED
+ FIELDS TERMINATED BY ':' collection items terminated by ',' location '${system:test.tmp.dir}/test_null_array';
```
Comment (ayushtkn) [inline ql/src/test/queries/clientpositive/udf_array_distinct.q:29]: This external table doesn't have purge ``true``, so drop won't delete the data, ${system:test.tmp.dir} won't get cleaned up after each query execution.
Better do LOAD data or if not possible do dfs -rm for all of the data folders/file that you have created

File: ql/src/test/queries/clientpositive/udf_array_distinct.q (hunk 6 above)
Comment (tarak271) [reply ql/src/test/queries/clientpositive/udf_array_distinct.q]: Added flag

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/AbstractGenericUDFArrayBase.java (hunk 5 above)
Comment (tarak271) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/AbstractGenericUDFArrayBase.java]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java (hunk 4 above)
Comment (tarak271) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java]: Changed as recommended

Comment (tarak271) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayDistinct.java]: And added another Test class to simulate null Object, TestGenericUDFArray which can be reused across Array functions

**Comment by tarak271** [comment]
Test failure seems unrelated to this changes. That test 'orc_ppd_basic.q' is even failing without my changes

**Review by ayushtkn — APPROVED** [review]
LGTM
